subsys/net/lib/openthread/platform/radio.c

```diff
@@ -1019,23 +1019,113 @@ uint8_t otPlatRadioGetCslAccuracy(otInstance *aInstance)
+ * | Length    | Element ID | Type=0 |      Vendor OUI      |
+ * +-----------+------------+--------+----------------------+
+ * |           Bytes: 0-1            |          2-4         |
+ * +-----------+---------------------+----------------------+
+ * | Bits: 0-6 |    7-14    |   15   | IE_VENDOR_THREAD_OUI |
+ * +-----------+------------+--------+----------------------|
+ *
+ * Thread v1.2.1 Spec., 4.11.3.4.4.6
+ * +---------------------------------+-------------------+------------------+
+ * |                  Vendor Specific Information                           |
+ * +---------------------------------+-------------------+------------------+
+ * |                5                |         6         |   7 (optional)   |
+ * +---------------------------------+-------------------+------------------+
+ * | IE_VENDOR_THREAD_ACK_PROBING_ID | LINK_METRIC_TOKEN | LINK_METRIC_TOKEN|
+ * |---------------------------------|-------------------|------------------|
+ */
+static uint8_t set_vendor_ie_header(bool lqi, bool link_margin, bool rssi, uint8_t *ie_header)
+{
+	/* Vendor-specific IE identifier */
+	const uint8_t ie_vendor_id = 0x00;
+	/* Thread Vendor-specific ACK Probing IE subtype ID */
+	const uint8_t ie_vendor_thread_ack_probing_id = 0x00;
+	/* Thread Vendor-specific IE OUI */
+	const uint32_t ie_vendor_thread_oui = 0xeab89b;
+	/* Thread Vendor-specific ACK Probing IE RSSI value placeholder */
+	const uint8_t ie_vendor_thread_rssi_token = 0x01;
+	/* Thread Vendor-specific ACK Probing IE Link margin value placeholder */
+	const uint8_t ie_vendor_thread_margin_token = 0x02;
+	/* Thread Vendor-specific ACK Probing IE LQI value placeholder */
+	const uint8_t ie_vendor_thread_lqi_token = 0x03;
+	const uint8_t ie_header_size = 2;
+	const uint8_t oui_size = 3;
+	const uint8_t sub_type = 1;
+	const uint8_t id_offset = 7;
+	const uint16_t id_mask = 0x00ff << id_offset;
+	const uint8_t type = 0x00;
+	const uint8_t type_offset = 7;
+	const uint8_t type_mask = 0x01 << type_offset;
+	const uint8_t length_mask = 0x7f;
+	uint8_t content_len;
+	uint16_t element_id = 0x0000;
+	uint8_t link_metrics_idx = 6;
+	uint8_t link_metric_data_len = (uint8_t)lqi + (uint8_t)link_margin + (uint8_t)rssi;
+
+	__ASSERT(link_metric_data_len <= 2, "Thread limits to 2 metrics at most");
+	__ASSERT(ie_header, "Invalid argument");
+
+	if (link_metric_data_len == 0) {
+		return 0;
+	}
+
+	/* Set Element ID */
+	element_id = (((uint16_t)ie_vendor_id) << id_offset) & id_mask;
+	sys_put_le16(element_id, &ie_header[0]);
+
+	/* Set Length - number of octets in content field. */
+	content_len = oui_size + sub_type + link_metric_data_len;
+	ie_header[0] = (ie_header[0] & ~length_mask) | (content_len & length_mask);
+
+	/* Set Type */
+	ie_header[1] = (ie_header[1] & ~type_mask) | (type & type_mask);
+
+	/* Set Vendor Oui */
+	sys_put_le24(ie_vendor_thread_oui, &ie_header[2]);
+
+	/* Set SubType */
+	ie_header[5] = ie_vendor_thread_ack_probing_id;
+
+	/* Set Link Metrics Tokens */
```

Comment:
Thread requires the order of requested metrics by the Link Metrics Initiator to be kept by the Link Metrics Subject in the ACKs. This is not yet implemented by OpenThread, but what about adding some TODO comment here?